### PR TITLE
[None][feat] Add chunked prefill support for Gemma4 (text + vision multimodal)

### PR DESCRIPTION
**Pull request by Hudayday**
## Summary

  - Add chunked prefill support for Gemma4 (all 4 variants: E2B, E4B, 26B-A4B MoE, 31B) for both text-only and multimodal (image + audio) requests
  - Builds on top of #12932 (Gemma4 base support). Independent of any KV-cache-manager-V2 leak fixes — those will land in a separate follow-up PR
  - **SWA chunked prefill via Triton**: trtllm-gen has no `head_dim=512` + sliding-window + chunked-prefill cubin, so we route SWA-layer prefill through the existing Triton kernel (added in #12932) once a request is in chunked mode. Causal mask
  path supported (no custom mask needed)
  - **MM bidirectional block chunk alignment**: Modified V2 scheduler aligns each chunk boundary to bidirectional MM block boundaries (Gemma4 26B/31B vision blocks), so a vision block never gets split across two prefill chunks (which would break the
  bidirectional attention mask invariant)
  - **Companion `triton_prefill` int64-stride fix**: 26B/31B production trigger — `safe_page_ids * stride_buf_kpage` was int32 and overflowed once `page_id * stride > 2^31`, producing illegal memory accesses on B200 chunked MMMU. Single
  `.to(tl.int64)` cast covers the offset arithmetic.
  - **Validated end-to-end**: smoke MMMU (image) + GSM8K (text-only) across all 4 variants on B200 single-GPU, plus full-set 31B mmmu (900 samples) — all green

  ## Modified Files

  | File | Purpose |
  |------|---------|
  | `_torch/attention_backend/triton_prefill.py` | Routing knob: enter Triton path on SWA + chunked + no-custom-mask. `safe_page_ids.to(tl.int64)` overflow fix. |
  | `_torch/attention_backend/flashinfer.py` | Gate Triton-fallback selection from FlashInfer backend in chunked mode. |
  | `_torch/models/modeling_gemma4.py` | Wire chunked-prefill path through Gemma4Attention for SWA layers. |
  | `_torch/models/modeling_gemma4mm.py` | Per-instance `mm_bidirectional_blocks` flag derived from `text_config.use_bidirectional_attention == "vision"`. `get_mm_token_ids` override. Audio `get_num_tokens_per_audio` override. |
  | `_torch/models/modeling_multimodal_utils.py` | Slice `mm_embeds` per chunk before `fuse_input_embeds`. Drop no-op log. |
  | `_torch/pyexecutor/scheduler/scheduler_v2.py` | `_align_chunk_to_mm_block` — INCLUSIVE-cumsum chunk-to-bidirectional-block alignment, with impossibility raise when MM block > `max_num_tokens`. |
  | `inputs/registry.py` | `setdefault(mm_bidirectional_blocks)` before cumsum idempotency guard so the hashing path also populates the gate. `isinstance(dict)` Mock-truthy guard. |
  | `tests/unittest/_torch/executor/test_kv_cache_v2_scheduler.py` | Unit tests for `_align_chunk_to_mm_block` (INCLUSIVE cumsum convention, MM-block-split scenarios, impossibility raise, snap-down test fix). |

  ## Key Design Decisions

  - **Triton-prefill routing**: Re-use the existing Triton prefill kernel added in #12932 (originally for `head_dim=512` + custom mask). Chunked-prefill SWA layers also enter this kernel, but with `custom_mask=None` (auto-generated causal). This
   avoids requiring a new trtllm-gen cubin for `H512` + chunked + SWA.
  - **MM chunk alignment — INCLUSIVE cumsum**: The chunk aligner consumes `embed_mask.cumsum(0)` of length `P` written by `registry.py` / runtime / `ad_executor`. The scheduler now reads it as INCLUSIVE (`mask[i] = cumsum[i] - cumsum[i-1]`)
  instead of the SHIFTED `[P+1]` convention, eliminating the out-of-range `cumsum[end_abs+1]` read that previously tripped `IndexError @ scheduler_v2.py:572` when the last chunk landed on `prompt_len-1`.
  - **Per-instance `mm_bidirectional_blocks`**: Set only when `text_config.use_bidirectional_attention == "vision"` (26B/31B). E2B/E4B (fully causal text + audio-only MM) skip the aligner entirely — their large audio soft-token block (≈451)
  would otherwise exceed `max_num_tokens` and trip the impossibility raise.
  - **MM embed slicing per chunk**: `fuse_input_embeds` receives only the slice of `mm_embeds` covering the current chunk's positions, so the mm-embed-to-text-positions alignment stays valid across chunks.

  ## Companion Bug Fixes

  - **`triton_prefill` int32 overflow** — `page_id * stride` silently wrapped to negative for `page_id > 16K` on a `[N, 2, 8, 32, 256]` KV layout (stride 131072). Deterministic IMA on first `prefix_lens > 0` iter. Single `.to(tl.int64)` cast on
  `safe_page_ids` fixes both K and V offset arithmetic.
  - **Audio hashing path** — Gemma4 inherited `get_num_tokens_per_image/_video` but had no `get_num_tokens_per_audio`. After image-then-audio request sequences locked `multimodal_hashing_supported=True`, audio raised `AttributeError`. Added
  override returning `processor.audio_seq_length` (750 fixed for E2B/E4B).
  - **Mock-truthy `py_multimodal_data` gate** — was `getattr-or-{}`, which doesn't short-circuit on `Mock`. Changed to `isinstance(data, dict)`.

  ## Test Plan

  - `test_kv_cache_v2_scheduler` unit tests (INCLUSIVE-cumsum chunk-to-mm-block alignment, impossibility raise, snap-down preempted, multi-block scenarios) — **all green locally** (153 V2 + 16 MM tests).
  - B200 single-GPU MMMU smoke (NUM_SAMPLES=30, `max_num_tokens=2048`, `max_seq_len=16384`) for **all 4 variants** (E2B / E4B / 26B / 31B) — all green.
  - B200 single-GPU GSM8K smoke (NUM_SAMPLES=50, text-only) for **all 4 variants** — all green.
  - B200 single-GPU MMMU full set (NUM_SAMPLES=900, `max_num_tokens=4096`, `max_seq_len=16384`) on 31B chunked + nochunk — accuracy matches non-chunked baseline within seed noise.

  ## Known Limitations

  - KV-cache-manager-V2 high-pressure leak (overlap-scheduler + suspend-as-victim desync) is **out of scope** for this PR. Validation here is on configurations where that leak does not trigger, or with the V2-leak fix applied externally.
  - **Multimodal towers are still HF-based.** Vision tower and audio tower in `modeling_gemma4mm.py` use `AutoModel.from_config(config.vision_config)` / `AutoModel.from_config(config.audio_config)` from `transformers`, carried over from #12932.
  Per [@yechank-nvidia's review on #12932](https://github.com/NVIDIA/TensorRT-LLM/pull/12932):
    > Can you re-write the vision_tower into our TRT-LLM's style? We deprecated all direct HF vision tower import.

    As discussed in that thread, the tower port is deferred to a follow-up PR because it requires (a) a `SiglipVisionModel`-equivalent TRT-LLM module, (b) a custom weight remapper for the multimodal checkpoint, and (c) a Siglip2-config compat
  check against the existing `modeling_siglip.py`. The audio tower has the same shape and will ship in the same follow-up.

## Summary by CodeRabbit

* **New Features**
  * Added sliding-window constraint support for Triton prefill attention operations.
  * Enhanced multimodal attention to respect bidirectional block boundaries during request scheduling.
  * Improved Gemma4 support for cached/paged prefix concatenation with multimodal tokens.
  * Added audio token counting predictions for Gemma4 multimodal processing.

* **Bug Fixes**
  * Hardened out-of-window KV cache page handling in Triton prefill.

* **Tests**
  * Added comprehensive test suite for multimodal-aware chunk alignment validation.

[Review Change Stack](https://app.coderabbit.ai/change-stack/NVIDIA/TensorRT-LLM/pull/14134)

### CODE REVIEW

File: tensorrt_llm/inputs/registry.py [hunk 1]
```diff
@@ -136,6 +136,14 @@ class BaseMultimodalInputProcessor(ABC):
     processes the multimodal inputs.
     """
 
+    # Whether multimodal soft-token runs need bidirectional attention spanning
+    # the full block. When True, the chunked-prefill scheduler must keep each
+    # MM block intact within a single iteration (snap-up or snap-down at chunk
+    # boundaries) so the second half does not attend to a frozen first half.
+    # Default False: most VLMs (Llava, Qwen-VL) use causal attention over MM
+    # tokens and tolerate splitting. Gemma4 sets True.
+    mm_bidirectional_blocks: ClassVar[bool] = False
```
Comment (2ez4bz) [inline tensorrt_llm/inputs/registry.py:145]: I might be missing something - why do we need this here if the check is on the instance at line 840?

File: tensorrt_llm/_torch/models/modeling_gemma4mm.py [hunk 2]
```diff
@@ -273,6 +290,39 @@ def get_vocab_size(self) -> Optional[int]:
...
+        # zero MM positions, producing an all-zero embed_mask cumsum and
+        # making chunked-prefill drop every MM embedding row. HF's
+        # `Gemma4Processor` does not expose `mm_token_ids` either, so we
+        # build the list from the config here.
+        ids = []
+        for name in ("image_token_id", "audio_token_id", "video_token_id"):
+            tid = getattr(self._config, name, None)
+            if tid is not None:
+                ids.append(int(tid))
+        if not ids:
```
Comment (2ez4bz) [inline tensorrt_llm/_torch/models/modeling_gemma4mm.py:306]: Based on the code comment, wouldn't this be wrong? If so, should we raise an error instead?

File: tensorrt_llm/_torch/models/modeling_multimodal_utils.py [hunk 3]
```diff
@@ -321,8 +318,10 @@ def find_input_mm_embeds(
 
     if len(mm_embeds) == 1:
         sliced = [mm_embeds[0][start:end] for start, end in slices]
-        return [torch.cat(sliced, dim=0)]
-    return [mm_embeds[i][start:end] for i, (start, end) in enumerate(slices)]
+        out = [torch.cat(sliced, dim=0)]
```
Comment (2ez4bz) [inline tensorrt_llm/_torch/models/modeling_multimodal_utils.py]: Were these just cosmetic changes?

**Review by 2ez4bz — APPROVED** [review]
Approving changes for the modeling files.

File: tensorrt_llm/_torch/models/modeling_gemma4mm.py (hunk 2 above)
Comment (Hudayday) [reply tensorrt_llm/_torch/models/modeling_gemma4mm.py:306]: Yes, we should raise an error here. Thanks for the suggestion, fixed in the following commit.

File: tensorrt_llm/_torch/models/modeling_multimodal_utils.py (hunk 3 above)
Comment (Hudayday) [reply tensorrt_llm/_torch/models/modeling_multimodal_utils.py]: Yes, those were cosmetic which I added when I was debugging. Reverted `tensorrt_llm/_torch/models/modeling_multimodal_utils.py` to the original.

File: tensorrt_llm/inputs/registry.py (hunk 1 above)
Comment (Hudayday) [reply tensorrt_llm/inputs/registry.py:145]: It has been added here as a template for any MM processors. 
Putting the default on the base class lets any subclass with a static value opt in with one line (`mm_bidirectional_blocks = True`) and skip `__init__` entirely, and keeping it on the base standardizes the contract so future MM processors don't have to rediscover the flag.

Gemma4 uses the same flag and just additionally sets it per-instance in `__init__` because the value depends on `text_config.use_bidirectional_attention`.

**Review by lfr-0531 — APPROVED** [review]

**Review by lancelly — APPROVED** [review]
Scheduler changes looks good to me

**Review by PerkzZheng — APPROVED** [review]
